e2e/internal/remote/remote.go

```diff
@@ -21,41 +21,57 @@ import (
 // commandTimeout is the maximum time a command can run before being cancelled.
 const commandTimeout = 90 * time.Minute
 
+// DomainUserPassword is the password to login as domain users.
+const DomainUserPassword = "supersecretpassword"
+
 // Client represents a remote SSH client.
 type Client struct {
-	*ssh.Client
+	client *ssh.Client
+	config *ssh.ClientConfig
+	host   string
 }
 
 // NewClient creates a new SSH client.
-// It parses the private key from the given path and establishes a connection to
-// the remote host.
-func NewClient(host string, username string, sshKeyPath string) (Client, error) {
-	privateBytes, err := os.ReadFile(sshKeyPath)
-	if err != nil {
-		return Client{}, err
-	}
-
-	signer, err := ssh.ParsePrivateKey(privateBytes)
-	if err != nil {
-		return Client{}, err
+// It establishes a connection to the remote host using the given authentication.
+// The secret will be treated as a private key if the path exists, otherwise it
+// will be treated as a password.
+func NewClient(host string, username string, secret string) (Client, error) {
+	var authMethod ssh.AuthMethod
+	privateBytes, err := os.ReadFile(secret)
+	if err == nil {
+		signer, err := ssh.ParsePrivateKey(privateBytes)
+		if err != nil {
+			return Client{}, err
+		}
+		authMethod = ssh.PublicKeys(signer)
+	} else {
+		// Could not read file, assuming password authentication
+		authMethod = ssh.Password(secret)
 	}
 
 	config := &ssh.ClientConfig{
 		User: username,
-		Auth: []ssh.AuthMethod{
-			ssh.PublicKeys(signer),
-		},
+		Auth: []ssh.AuthMethod{authMethod},
 		// nolint:gosec // This is used for E2E tests where machines are created on the fly
 		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
 		Timeout:         10 * time.Second,
 	}
 
 	client, err := ssh.Dial("tcp", host+":22", config)
 	if err != nil {
-		log.Fatalf("Failed to dial: %v", err)
+		return Client{}, fmt.Errorf("failed to establish connection to remote host: %w", err)
 	}
 
-	return Client{client}, nil
+	return Client{
+		client: client,
+		config: config,
+		host:   host,
+	}, nil
+}
+
+// Close closes the SSH connection.
+func (c Client) Close() error {
```

Comment:
I would put a pointer on Client here, just to signal we change something on the underlying object (even if it’s a copy of an object handling a pointer has no special effect, but at least, semantically, we signal it correct)